auth-service/src/main/kotlin/uk/gov/communities/delta/auth/controllers/external/DeltaSetPasswordController.kt

```diff
@@ -0,0 +1,199 @@
+                "Passwords must not be a commonly used password format or contain your username"
+            else null
+
+        val errorPresent = message != null
+        val tokenResult = setPasswordTokenService.useToken(token, userCN, !errorPresent)
+        call.respondToResult(tokenResult, message, newPassword)
+    }
+
+    private suspend fun sendNewSetPasswordLink(userCN: String) {
+        val user = userLookupService.lookupUserByCn(userCN)
+        emailService.sendTemplateEmail(
+            "not-yet-enabled-user",
+            EmailContacts(
+                user.email!!,
+                user.fullName,
+                emailConfig.fromEmailAddress,
+                emailConfig.fromEmailName,
+                emailConfig.replyToEmailAddress,
+                emailConfig.replyToEmailName,
+            ),
+            "DLUHC DELTA - Set Your Password",
+            mapOf(
+                "deltaUrl" to deltaConfig.deltaWebsiteUrl,
+                "userFirstName" to user.firstName,
+                "setPasswordUrl" to getSetPasswordURL(
+                    setPasswordTokenService.createToken(userCN),
+                    userCN,
+                    authServiceConfig.serviceUrl
+                )
+            )
+        )
+    }
+
+    class SetPasswordException(
+        errorCode: String,
+        exceptionMessage: String,
+        userVisibleMessage: String = "Something went wrong, please click the link in your account activation email again",
+    ) : UserVisibleServerError(errorCode, exceptionMessage, userVisibleMessage, "Set Password Error")
+
+    private suspend fun ApplicationCall.respondToResult(
+        tokenResult: SetPasswordTokenService.TokenResult,
+        message: String?,
+        newPassword: String
+    ) {
+        when (tokenResult) {
+            is SetPasswordTokenService.NoSuchToken -> {
+                throw SetPasswordException(
+                    "set_password_invalid_token",
+                    "Token did not exist on setting password"
+                )
+            }
+
+            is SetPasswordTokenService.ExpiredToken -> {
+                sendNewSetPasswordLink(tokenResult.userCN)
+                this.respondExpiredTokenPage()
+            }
+
+            is SetPasswordTokenService.ValidToken -> {
+                if (message != null) return this.respondSetPasswordPage(tokenResult.userCN, tokenResult.token, message)
+                val userDN = String.format(ldapConfig.deltaUserDnFormat, tokenResult.userCN)
+                try {
+                    userService.setPassword(userDN, newPassword)
+                } catch (e: Exception) {
+                    logger.error("Error setting password for user with DN {}", userDN, e)
+                    return this.respondSetPasswordPage(
+                        tokenResult.userCN,
+                        tokenResult.token,
+                        "An error occurred please try again or contact the service desk"
+                    )
```

Comment:
I think rethrow and use the generic error page here, that way we get a request id